mt32emu_qt/src/audiodrv/CoreAudioDriver.cpp

```diff
@@ -143,7 +157,59 @@ CoreAudioDriver::~CoreAudioDriver() {
 
 const QList<const AudioDevice *> CoreAudioDriver::createDeviceList() {
 	QList<const AudioDevice *> deviceList;
-	deviceList.append(new CoreAudioDevice(*this));
+	deviceList.append(new CoreAudioDevice(*this)); // default device
+    
+    // Get system output devices
+    UInt32 propertySize = 0;
+    int numDevices = 0;
+    QVector<AudioDeviceID> deviceIDs;
+    AudioObjectPropertyAddress propertyAddress;
+    propertyAddress.mSelector = kAudioHardwarePropertyDevices;
+    propertyAddress.mScope = kAudioObjectPropertyScopeGlobal;
+    propertyAddress.mElement = kAudioObjectPropertyElementMaster;
+    
+    if (AudioObjectGetPropertyDataSize(kAudioObjectSystemObject, &propertyAddress, 0, 0, &propertySize) == noErr) {
+        numDevices = propertySize / sizeof(AudioDeviceID);
+        deviceIDs.resize(numDevices);
+        
+        if (AudioObjectGetPropertyData(kAudioObjectSystemObject, &propertyAddress, 0, NULL, &propertySize, deviceIDs.data()) == noErr) {
+            foreach (const AudioDeviceID &id, deviceIDs) {
+                AudioObjectPropertyAddress deviceAddress;
+                propertySize = 0;
+                deviceAddress.mSelector = kAudioDevicePropertyStreams;
+                deviceAddress.mScope = kAudioObjectPropertyScopeOutput;
+                deviceAddress.mElement = kAudioObjectPropertyElementMaster;
+                
+                if (AudioObjectGetPropertyDataSize(id, &deviceAddress, 0, NULL, &propertySize) == noErr) {
+                    if (propertySize > 0) {
+                        CFStringRef devUidRef;
+                        propertySize = sizeof(CFStringRef);
+                        deviceAddress.mSelector = kAudioDevicePropertyDeviceUID;
+                        deviceAddress.mScope = kAudioObjectPropertyScopeGlobal;
+                        deviceAddress.mElement = kAudioObjectPropertyElementMaster;
+                        
+                        if (AudioObjectGetPropertyData(id, &deviceAddress, 0, NULL, &propertySize, &devUidRef) == noErr) {
+                            CFStringRef devNameRef;
+                            propertySize = sizeof(CFStringRef);
+                            deviceAddress.mSelector = kAudioDevicePropertyDeviceNameCFString;
+                            deviceAddress.mScope = kAudioObjectPropertyScopeGlobal;
+                            deviceAddress.mElement = kAudioObjectPropertyElementMaster;
+                            
+                            if (AudioObjectGetPropertyData(id, &deviceAddress, 0, NULL, &propertySize, &devNameRef) == noErr) {
+                                QString uid, name;
+                                uid = QString::fromCFString(devUidRef);
+                                name = QString::fromCFString(devNameRef);
```

Comment:
According to Qt docs, this function was introduced in Qt 5.2. Is there a way to make it compilable with older versions (since Qt4 remains supported)?

Reply:
I copied the code for converting to QString from CFString from Qt 5.2 source code, is this acceptable?

Reply:
Yes, I'd take it as an acceptable approach since the Qt code is covered by a compatible license. But it's a little surprising to me that there was no such a method before, hmm.

Reply:
Yes, It confuses me a little why did they put that method in a private class in previous versions